[ADP-3272] Add the `TxWithUTxO` data type.

jonathanknowles opened the pull request:
## Summary

This PR:

- adds the `TxWithUTxO` data type, along with smart constructors, generators, and shrinkers.
- uses the `TxWithUTxO` data type to simplify the `Arbitrary` instance for `PartialTx`.

## Details

An object of type `TxWithUTxO` combines a transaction `t` with an associated UTxO set `u`, such that:
- every input in transaction `t` is **_guaranteed_** to resolve to a UTxO within UTxO set `u`. 
- UTXO set `u` **_may_** also contain **_additional_** UTxOs that are not referenced by `t`.

In order to uphold the above guarantee:
- the `TxWithUTxO` data type is defined within its own module;
- the `TxWithUTxO` data constructor is not exported;
- a pattern synonym `TxWithUTxO` is provided to facilitate safe pattern matching.

## Motivation

This data type will be used for testing `balanceTx` in situations where we want to guarantee that `balanceTx` will not fail with `ErrBalanceTxUnresolvedInputs`.

## Issue

ADP-3272

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO.hs:
```diff
@@ -0,0 +1,101 @@
+-- | Constructs a 'TxWithUTxO' object.
+--
+-- The given transaction will be stripped of any inputs that do not resolve to
+-- UTxOs with the given UTxO set.
```
🤔 Why not fail instead of silently stripping?

jonathanknowles replied:
> Why not fail instead of silently stripping?

That's a fair point!

I think the ability to strip unresolved inputs is useful for building shrinkers, so we do want to have that ability **_somewhere_**.

On reflection though, perhaps this shouldn't be the **_default_** behaviour.

I've made the following changes:
1. renamed `construct` to `constructFilter`;
1. added a second constructor called `constructEither` that will fail if there are any unresolved inputs;
1. removed the constructor pattern synonym.

Anviking replied:
`constructFilter` — even with that name — seems very dangerous to me. Are there any drawbacks to moving it to the `Gen` module? Being required to `fromJust` the result of `constructFilter` instead of having access to `UnsafeTxWithUTxO` seems like a good thing to me if we did that. If we'd forget to update `removeUnresolvableInputs`, then `unresolvableInputs`'s `allInputsTxBodyF` would catch it and cause a failure.

We could also name it more in the style of `accursedUnutterablePerformIO` — e.g. `construct` and `constructBySilentlyDroppingUnresolvableInputs`... But moving it would seem safer / saner to me.

(I guess `TxWithUTxO` isn't a public module... but then again, if we want to expose it, creaing a public version containing everything but `constructFilter` seems very overkill to me)

jonathanknowles replied:
I've renamed these to:

```hs
construct         :: Tx era -> UTxO era -> Either (NESet TxIn) (TxWithUTxO era)
constructFiltered :: Tx era -> UTxO era ->                      TxWithUTxO era
```

- The `construct` function only succeeds if all inputs are resolvable. Otherwise, it returns a non-empty set of unresolvable inputs.
- The `constructFiltered` function always succeeds by filtering out any inputs from the transaction that are not resolvable.

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO.hs:
```diff
@@ -0,0 +1,101 @@
+
+    removeUnresolvableInputs :: TxBody era -> TxBody era
+    removeUnresolvableInputs
+        = over           inputsTxBodyL f
+        . over collateralInputsTxBodyL f
```
Notably missing: `referenceInputsTxBodyL` 😄 

It would be nice to use `allInputsTxBodyF` https://github.com/IntersectMBO/cardano-ledger/blob/71efc6b19fd70df3e2c4e639b89f30b641e9d62a/libs/cardano-ledger-core/src/Cardano/Ledger/Core.hs#L212-L215 (reexported in `Cardano.Ledger.Api.Tx.Body`), although I see it wouldn't work with the `over` filtering approach, as it's just a getter.

jonathanknowles replied:
> Notably missing: referenceInputsTxBodyL 😄

Thanks! I've now added this.

> It would be nice to use allInputsTxBodyF

Good point. I've used `allInputsTxBodyF` in the definition of `unresolvedInputs`.

Anviking reviewed (commented):
First look

jonathanknowles committed:
Add dependency on package `either`.

jonathanknowles commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -467,6 +471,7 @@ import qualified Data.Text.IO as T
 import qualified Internal.Cardano.Write.Tx as Write
 import qualified Ouroboros.Consensus.HardFork.History as HF
 import qualified Test.Hspec.Extra as Hspec
+import Internal.Cardano.Write.Tx.TxWithUTxO (pattern TxWithUTxO, type TxWithUTxO)
```
Wrong place.

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO/Gen.hs:
```diff
@@ -0,0 +1,115 @@
+
+-- | Generates a 'TxWithUTxO' object that has a non-minimal UTxO set.
+--
+-- The domain of the UTxO map is a strict superset of the transaction input set.
```
```suggestion
-- The domain of the UTxO map is a superset of the transaction input set.
```

Couldn't `surplusUTxO` be a subset of `minimalUTxO` such that `minimalUTxO <> surplusUTxO == minimalUTxO`? (With the right generator and seed)

jonathanknowles replied:
@Anviking I've changed the generator to guarantee that the result is a strict superset.

Anviking commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -2204,24 +2212,24 @@ instance Arbitrary (MixedSign Value) where
-        | tx' <- shrinkTx tx
-        ]
+    arbitrary =
+        mkPartialTx <$> TxWithUTxO.generate genTxForBalancing genTxIn genTxOut
```
`50%` with surplus UTxO set seems unnecessary... why not <s>100% or 90%</s> 0 or 10%?

jonathanknowles replied:
@Anviking With the current `Arbitrary` instance for `PartialTx`, the proportion of generated `PartialTx` objects with surplus UTxO sets is **0%**.

This is because the domain of the generated `extraUTxO` is precisely equal to the set of inputs in the transaction:

```hs
instance forall era. IsRecentEra era => Arbitrary (PartialTx era) where
    arbitrary = do
        tx <- genTxForBalancing
        extraUTxO <- genExtraUTxO (txInputs tx)
        pure PartialTx {tx, extraUTxO, redeemers, timelockKeyWitnessCounts}
      where
        genExtraUTxO :: Set TxIn -> Gen (UTxO era)
        genExtraUTxO = fmap UTxO . genMapFromKeysWith genTxOut
        txInputs :: Tx era -> Set TxIn
        txInputs tx = tx ^. bodyTxL . inputsTxBodyL
```

But as we know, the semantics of `balanceTx` do allow for a `PartialTx` to have surplus UTxOs not referenced by the inputs of the inner `Tx`.

This PR changes the ratio from **0%** to **50%**, so that we are now (at least some of the time) testing the case where there is a UTxO surplus within the `PartialTx` UTxO set.

Anviking replied:
Apologies - I meant the opposite 😅 : Why not e.g. 10% (going from 0% to 10%)? Most tests in `BalanceSpec` shouldn't care about the possibility for the extra UTxO set to contain a surplus to what is needed.

jonathanknowles replied:
@Anviking Fair enough!

I've adjusted the frequency distribution in this commit:
403ed7df664637b0041c7dce5bd5c92d1a3ec1e8

Anviking replied:
👍 I believe this brought the success rate in the "produces valid transactions or fails" property from 2% back up to 4% (which btw still was very low 👀 )

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO/Gen.hs:
```diff
@@ -0,0 +1,115 @@
+    utxo <- UTxO <$> genMapFromKeysWith genTxOut (txInputs tx)
+    pure $ TxWithUTxO.constructFiltered tx utxo
+  where
+    txInputs = view (bodyTxL . allInputsTxBodyF)
```
IMO `all` is a key detail and good to have in the name if not inlining it. E.g. `allTxInputs`

jonathanknowles replied:
I've changed it to `txAllInputs`.

jonathanknowles committed:
Remove unused function `restrictResolution`.

jonathanknowles committed:
Remove unused function `shrinkInputResolution`.

jonathanknowles committed:
Add dependency on package `cardano-wallet-test-utils`.

jonathanknowles committed:
Add module `Cardano.Write.Tx.TxWithUTxO`.

jonathanknowles committed:
Add module `Cardano.Write.Tx.TxWithUTxO.Gen`.

jonathanknowles committed:
Redefine `Arbitrary` instance for `PartialTx`.

Anviking commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -2203,25 +2198,29 @@ instance Arbitrary (MixedSign Value) where
+shrinkTxWithUTxO = TxWithUTxO.shrinkWith shrinkTx shrinkUTxOToSubsets
+  where
+    shrinkUTxOToSubsets :: IsRecentEra era => UTxO era -> [UTxO era]
+    shrinkUTxOToSubsets = shrinkMapBy UTxO unUTxO shrinkMapToSubmaps
```
Could you move these to `where` clauses of the `Arbitrary (PartialTx era)`? I think it would be cleaner (wrt module name space) and more clear where they're used.

jonathanknowles replied:
Hi @Anviking 

It's not possible to move `mkPartialTx`, because it is used by both `arbitrary` and `shrink`.

AFAIK it's not possible to have a `where` clause that is common to the entire instance definition.

jonathanknowles replied:
Though I have renamed these functions in: 5262ce259cb9db4c4d6366be0bcaafa134138b0c.

jonathanknowles committed:
Adjust the frequency distribution in `TxWithUTxO.Gen.generate`.

In response to review feedback:
https://github.com/cardano-foundation/cardano-wallet/pull/4570#discussion_r1589650350

jonathanknowles committed:
Rename functions `mkPartialTx` and `unPartialTx`.

We use the following new names:

- `partialTxFromUTxOWithTx`
- `UTxOWithTxFromPartialTx`

jonathanknowles committed:
Draw more attention to constants within `partialTxFromTxWithUTxO`.

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO/Gen.hs:
```diff
@@ -0,0 +1,130 @@
+shrinkWith shrinkTx shrinkUTxO txWithUTxO =
+    interleaveRoundRobin
+        [ shrinkTxWith   shrinkTx   txWithUTxO
+        , shrinkUTxOWith shrinkUTxO txWithUTxO
```
🤔 Looking at this I came to think: shouldn't an early shrinking step be dropping all surperflous UTxOs? And then: would there be a case to make for having `TxWithUTxO` _never_ contain surperflous UTxO entries? The only thing they affect would be the utxo inconsistency check.

Would depend on what your plans are, but perhaps food for thought for later.

Anviking commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -2203,25 +2198,36 @@ instance Arbitrary (MixedSign Value) where
+    timelockKeyWitnessCounts = mempty
+
+txWithUTxOFromPartialTx :: IsRecentEra era => PartialTx era -> TxWithUTxO era
+txWithUTxOFromPartialTx PartialTx {tx, extraUTxO} =
```
Could also write:

```suggestion
txWithUTxOFromPartialTx (PartialTx tx extraUTxO _redeemers _wits) =
```

to make it explicit that we are forgetting some information.

jonathanknowles replied:
This could also be done without sacrificing the use of named field puns. For example:

```hs
  txWithUTxOFromPartialTx
      PartialTx
          { tx
          , extraUTxO
          , redeemers                = _forgetThisField0
          , timelockKeyWitnessCounts = _forgetThisField1
          } =
          TxWithUTxO.constructFiltered
              tx
              extraUTxO
```

Although I'm not sure if this is really worth it, I do see your point.

In any case, I suspect we'll revisit this function definition very soon, so we can revise if necessary.

Anviking replied:
Ah, but also: with "positional patterns" we guarantee that all fields have to be listed. I.e. we seeing `(PartialTx tx extraUTxO _redeemers _wits)` we can trust that it's only ignoring `_redeemers` and `_wits`, and that no one has added a new field `foo` to `PartialTx` without concidering whether it should be ignored or not in this function.

I think this would be my one concern in general with named field puns.

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO/Gen.hs:
```diff
@@ -0,0 +1,130 @@
+generateWithMinimalUTxO
+    :: IsRecentEra era
+    => Gen (Tx era)
+    -> Gen (TxIn)
```
```suggestion
```

🤔 There's no point in keeping an unused `_genTxIn` for symmetry?

jonathanknowles replied:
@Anviking Fair enough!

I've removed this parameter in 71610da7baa0731cedc935f75ed4f34ee696ed7b.

Anviking reviewed (approved):
Thanks! Lgtm.

jonathanknowles committed:
Remove unused `Gen TxIn` argument from `generateWithMinimalUTxO`.

In response to review feedback:

https://github.com/cardano-foundation/cardano-wallet/pull/4570#discussion_r1596996702

jonathanknowles committed:
Make `genTxIn` a top-level function.

The new location is next to the other general-purpose generators, such
as `genTxOut`.
